### PR TITLE
Improve timer

### PR DESCRIPTION
**Pull request by holmanb**
## Note to reviewers

This looks big, but it is broken into three logically separate changes. I would recommend reviewing a commit at a time.

## Motivation

The recent addition of subp execution time was a huge benefit to exposing slow external commands. As we try to push performance more, more data is vital to making the biggest impact that we can. Comparing this branch against the tip of main, this change makes the total log size 1 line longer - but this will vary depending per-platform.

- builtin timing instrumentation in production allows diagnosing everywhere we can get a log
- reduce noise by only logging slow events
- avoid the need for manual instrumentation
- `cloud-init analyze` is neat, but it is low granularity, relies log timestamps (so is only an approximation), and it requires extra steps 
- standardize logging to in shared code

## Description of changes

- Improve on the existing timer helper by simplifying its usage and implementation
- Remove obsolete functionality in the existing timer function
- Eliminate irrelevant logs by only logging if a threshold is exceeded.
- Update existing callsites to use context manager
- Refactor logging helpers out of util.py
- Replace duplicate timer implementations with timer helper
  - socket.py
  - subp.py
- Add timer helper to shared callsites
  - DataSource.get_data()
  - each cloud-config module

## Todo

- fix unittests
- bikeshed the name (I originally implemented the context handler as a function, but now it's a class so it's suddenly un-pythonic)

## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: cloudinit/sources/helpers/azure.py [hunk 1]
```diff
@@ -457,7 +457,9 @@ def generate_certificate(self):
                 ]
             )
             certificate = ""
-            for line in open(self.certificate_names["certificate"]):
+            for line in util.load_text_file(
+                self.certificate_names["certificate"]
+            ).split("\n"):
```
Comment (TheRealFalcon) [inline cloudinit/sources/helpers/azure.py:462]: nit: .splitlines()

File: cloudinit/sources/__init__.py [hunk 2]
```diff
@@ -480,18 +488,19 @@ def get_data(self) -> bool:
...
-        self._dirty_cache = True
-        return_value = self._check_and_get_data()
-        # TODO: verify that datasource types are what they are expected to be
-        # each datasource uses different logic to get userdata, metadata, etc
-        # and then the rest of the codebase assumes the types of this data
-        # it would be prudent to have a type check here that warns, when the
-        # datatype is incorrect, rather than assuming types and throwing
-        # exceptions later if/when they get used incorrectly.
-        if not return_value:
+        with log.log_time("Getting metadata"):
```
Comment (TheRealFalcon) [inline cloudinit/sources/__init__.py:491]: This is a good addition, but I do wonder about in child classes where we were measuring something narrow in `_get_data()` that now may be much broader. In particular, I think it's useful knowing the time it took to crawl an IMDS without the surrounding processing of _get_data(). I think it's fine as-is, but it might be something we need to further tweak in the future.

File: cloudinit/sources/DataSourceEc2.py [hunk 3]
```diff
@@ -508,21 +499,6 @@ def network_config(self):
             return None
 
         result = None
-        no_network_metadata_on_aws = bool(
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceEc2.py:511]: This looks like a fine change, but I think it should be its own commit or PR.

File: cloudinit/ssh_util.py [hunk 4]
```diff
@@ -546,10 +546,9 @@ def parse_ssh_config_map(fname):
 def _includes_dconf(fname: str) -> bool:
     if not os.path.isfile(fname):
         return False
-    with open(fname, "r") as f:
-        for line in f:
-            if line.startswith(f"Include {fname}.d/*.conf"):
-                return True
+    for line in util.load_text_file(fname).split("\n"):
```
Comment (TheRealFalcon) [inline cloudinit/ssh_util.py:549]: Nit: .splitlines()

File: tests/unittests/config/test_cc_package_update_upgrade_install.py [hunk 5]
```diff
@@ -156,7 +156,7 @@ def _new_subp(*args, **kwargs):
         assert m_subp.call_args_list[0] == mock.call(["apt-cache", "pkgnames"])
 
         for arg in ["apt-get", "install", "pkg1", "pkg2"]:
-            assert arg in m_subp.call_args_list[1][1]["args"]
+            assert arg in m_subp.call_args_list[1][0][0]["args"]
```
Comment (TheRealFalcon) [inline tests/unittests/config/test_cc_package_update_upgrade_install.py:159]: What has changed about the args list in these tests? The three similar changes in this file all break tests for me.

File: tests/unittests/config/test_cc_package_update_upgrade_install.py [hunk 6]
```diff
@@ -241,21 +241,16 @@ def test_error_apt(self, common_mocks):
...
-            if args and "apt-cache" in args[0]:
-                return SubpResult("pkg1", None)
-            if "args" in kwargs and "install" in kwargs["args"]:
-                raise subp.ProcessExecutionError(
-                    cmd=kwargs["args"],
-                    stdout="dontcare",
-                    stderr="E: Unable to locate package pkg1",
-                    exit_code=100,
-                )
+            raise subp.ProcessExecutionError(
```
Comment (TheRealFalcon) [inline tests/unittests/config/test_cc_package_update_upgrade_install.py:244]: While the end result is the same, this changes the call path and where we're raising exceptions. I want to get past the apt cache phase and fail particularly in the apt install path as that is known to raise exceptions. I'm open to better indicating that in this test, but I'm not sure why this change is being made here.

File: tests/unittests/config/test_cc_ubuntu_pro.py [hunk 7]
```diff
@@ -1154,13 +1154,11 @@ def test_happy_path(
         m_should_auto_attach.should_auto_attach.return_value.should_auto_attach = (  # noqa: E501
             should_auto_attach_value
         )
-        if expected_result is None:  # Pro API does respond
+        if expected_result is None:
+            # Pro API does respond
             assert should_auto_attach_value == _should_auto_attach(ua_section)
-            assert (
```
Comment (TheRealFalcon) [inline tests/unittests/config/test_cc_ubuntu_pro.py:1159]: Hmmm...I think this highlights a gap in functionality where we might wind up losing valuable logging messages that no longer take as much time to execute. Looking at the Pro module now, there's now no log anywhere to indicate that we're attempting to attach to Pro, and I'm assuming that's not what we want. I'm not sure I have a great solution other than looking through all of the changed call sites and using a threshold of 0 those we think we want to log regardless of the time taken.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Thanks for this! Overall a really good change. I left some comments inline.

File: cloudinit/log/__init__.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    CustomLoggerType,
+    LogExporter,
+    configure_root_logger,
+    flush_loggers,
+    reset_logging,
+    setup_basic_logging,
+    setup_logging,
+)
+
+__all__ = [
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/log/__init__.py:16]: > Is there a reason you want to prefer importing from packages over importing from modules?

Mostly it was to keep from having to modify too many imports / callsites in this PR, but I suppose I probably already touched most of them.

I'm fine with just importing from modules - I actually slightly prefer it for cleaner separation.

File: cloudinit/log/log_util.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def logexc(
+    log, msg, *args, log_level: int = logging.WARNING, exc_info=True
+) -> None:
+    log.log(log_level, msg, *args)
+    log.debug(msg, exc_info=exc_info, *args)
+
+
+class log_time:
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/log/log_util.py:16]: > name bikeshed: Timed? Past tense because https://peps.python.org/pep-0343/#examples 

I like that a lot. Thanks for the suggestion.

> and no mention of log because it won't necessarily always log...which actually makes me wonder if this is the right place for this class.

+1 Maybe we do this instead?
```python
from cloudinit.performance import Timed
```

File: cloudinit/log/log_util.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+    def __enter__(self):
+        self.start = time.monotonic()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        delta = time.monotonic() - self.start
+        if delta > self._threshold:
+            if not self._skip_log:
+                LOG.debug("%s took %0.3f seconds", self._msg, delta)
+            self.output = f"{self._msg} took {delta:.3f} seconds"
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/log/log_util.py:58]: That's reasonable, though I do like being able to search for `took` in the log.

File: cloudinit/util.py [hunk 11]
```diff
@@ -2324,37 +2270,38 @@ def write_file(
     @param group: The group to set on the file.
     """
 
-    if preserve_mode:
-        try:
-            mode = get_permissions(filename)
-        except OSError:
-            pass
+    with log.log_time(f"Writing file {filename}"):
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/util.py:2273]: > Another idea would be to also have the same functionality we use in the context manager as a decorator.

I like that idea.

File: tests/unittests/config/test_cc_package_update_upgrade_install.py (hunk 5 above)
Comment (holmanb) [reply tests/unittests/config/test_cc_package_update_upgrade_install.py:159]: I think this is one of the tests that's actually failing currently.

File: cloudinit/sources/__init__.py (hunk 2 above)
Comment (holmanb) [reply cloudinit/sources/__init__.py:491]: > I think it's useful knowing the time it took to crawl an IMDS without the surrounding processing of _get_data()

Agreed. I went back and forth on this, and a previous iteration actually added a new standard `crawl_metadata()` to the base class and adapted other datasources to track this seperate from `get_data()`'s time. Ultimately though, I went for a simpler approach because I think that in most cases any extra time is already captured in the log: any datasources that need to set up ephemeral networks already get that captured via `subp`'s timer, and the data manipulation (mostly just building dicts) is trivial and doesn't involve complex data processing.

While this may be much broader in some cases, it is also more standardized across platforms which I think makes it more valuable.

> it might be something we need to further tweak in the future.

+1 no objections to making improvements wherever needed

File: cloudinit/sources/DataSourceEc2.py (hunk 3 above)
Comment (holmanb) [reply cloudinit/sources/DataSourceEc2.py:511]: See https://github.com/canonical/cloud-init/pull/5714

File: tests/unittests/config/test_cc_ubuntu_pro.py (hunk 7 above)
Comment (holmanb) [reply tests/unittests/config/test_cc_ubuntu_pro.py:1159]: Good point. I reworked the context manager to support a parameter called `log_mode`, with three options: "always", "threshold", and "skip" (default is "threshold"). I audited the modified callsites and set "always" to those that I think need it. Let me know if there are any others that you think should also be modified.

File: cloudinit/util.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/util.py]: > > Another idea would be to also have the same functionality we use in the context manager as a decorator.
> 
> I like that idea.

I've added a new decorator `timed()` and replaced this and several of the other call sites with the decorator.

**Comment by holmanb** [comment]
@TheRealFalcon thanks for the review. I think that I've addressed all of your comments. Ready for re-review.

File: cloudinit/atomic_helper.py [hunk 12]
```diff
@@ -34,8 +42,15 @@ def b64e(source):
 def write_file(
     filename, content, mode=_DEF_PERMS, omode="wb", preserve_mode=False
 ):
-    # open filename in mode 'omode', write content, set permissions to 'mode'
+    """open filename in mode omode, write content, set permissions to mode"""
+
+    with performance.Timed(f"Writing {filename}"):
```
Comment (TheRealFalcon) [inline cloudinit/atomic_helper.py:47]: Can the decorator be used here now instead?

File: cloudinit/safeyaml.py [hunk 13]
```diff
@@ -261,12 +267,13 @@ def load_with_marks(blob) -> Tuple[Any, Dict[str, int]]:
...
-    return yaml.dump(
-        obj,
-        line_break="\n",
-        indent=4,
-        explicit_start=explicit_start,
-        explicit_end=explicit_end,
-        default_flow_style=False,
-        Dumper=(NoAliasSafeDumper if noalias else yaml.dumper.SafeDumper),
-    )
+    with performance.Timed("Dumping yaml"):
```
Comment (TheRealFalcon) [inline cloudinit/safeyaml.py:270]: Can the new decorator work here?

File: cloudinit/sources/__init__.py [hunk 14]
```diff
@@ -475,6 +483,7 @@ def clear_cached_attrs(self, attr_defaults=()):
         if not attr_defaults:
             self._dirty_cache = False
 
+    @performance.timed("Getting metadata")
```
Comment (TheRealFalcon) [inline cloudinit/sources/__init__.py:486]: This should probably log always

File: cloudinit/performance.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        ```
+    """
+
+    def __init__(
+        self,
+        msg: str,
+        /,
+        *,
+        threshold: float = 0.01,
+        log_mode: str = "threshold",
```
Comment (TheRealFalcon) [inline cloudinit/performance.py:43]: I don't love this parameter being a magic string, especially when it only takes two values. Is there a reason setting threshold to 0 isn't a valid way to signal to log always? If so, could this field be a boolean like `always_log` or something similar?

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
I left a few more comments, but everything is mostly minor.

I really like the new decorator.

File: cloudinit/performance.py (hunk 15 above)
Comment (holmanb) [reply cloudinit/performance.py:43]: There are technically three (see "skip" in the docstring), not two, but the behavior of the third is implicit, which should probably be made explicit for code clarity.

> Is there a reason setting threshold to 0 isn't a valid way to signal to log always?

It would work, but then you may end up with messages ending with `.. took 0.000 seconds to complete`, or other numbers which come in under the threshold value, which is useless. By having three options `"skip", "always", "threshold"`, we're able to maximize signal to noise ratio.

> I don't love this parameter being a magic string

Nor do I, but it seems like the Pythonic way to control the mode of a function which has more than two modes. Consider the arguments in `"".encode()`'s, and `b''.decode()`. An alternative that I considered briefly was to define enums, but that requires more names to be imported and I wasn't able to find any examples of this in the Python standard library so this approach seems preferred.

@TheRealFalcon Thoughts? If you don't mind keeping it as a string, I can make the `"skip"` logic explicit.

File: cloudinit/config/modules.py [hunk 16]
```diff
@@ -283,11 +290,15 @@ def _run_modules(self, mostly_mods: List[ModuleDetails]):
                             deprecated_version="23.2",
                         )
                         func_args.update({"log": LOG})
-                    ran, _r = cc.run(
-                        run_name, mod.handle, func_args, freq=freq
-                    )
+
+                    with performance.Timed("and", log_mode="skip") as timer:
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/config/modules.py:294]: > This still feels a bit hard to understand at first. What do you think about having this message be "" and then later `f"{run_name} ran successfully and took {timer.delta:.3f} seconds"`?
> 
> Not a blocking comment if you prefer it as-is.

Not opposed, this seems a bit more intuitive.

File: cloudinit/performance.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+        might produce this log:
+        ```
+            Configuring the network took 0.100 seconds
+        ```
+    """
+
+    def __init__(
+        self,
+        msg: str,
+        /,
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/performance.py:40]: Having a standard calling convention makes code more recognizable but I don't feel strongly about this. I recall it also made refactoring slightly more difficult that one of the calls used the form `msg=""`, but I can drop it if you'd rather not.

File: cloudinit/atomic_helper.py (hunk 12 above)
Comment (holmanb) [reply cloudinit/atomic_helper.py:47]: The only reason that I didn't is that I couldn't think of a clean way to still include the filename in the log, which I think is useful for this log.

File: cloudinit/config/cc_resizefs.py [hunk 18]
```diff
@@ -315,20 +315,11 @@ def handle(name: str, cfg: Config, cloud: Cloud, args: list) -> None:
         # Fork to a child that will run
         # the resize command
         util.fork_cb(
-            util.log_time,
-            logfunc=LOG.debug,
-            msg="backgrounded Resizing",
-            func=do_resize,
-            args=(resize_cmd,),
+            do_resize,
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/config/cc_resizefs.py:318]: Yes. The thought process here is that it loos like other logs will indicate whether forking happened, and `do_resize()` just calls `subp()` which already uses `Timed()`.

File: cloudinit/cmd/main.py [hunk 19]
```diff
@@ -1300,13 +1304,8 @@ def sub_main(args):
...
     with args.reporter:
-        retval = util.log_time(
-            logfunc=LOG.debug,
-            msg="cloud-init mode '%s'" % name,
-            get_uptime=True,
-            func=functor,
-            args=(name, args),
-        )
+        with performance.Timed(f"cloud-init stage: '{rname}'"):
+            retval = functor(name, args)
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/cmd/main.py:1308]: This was a hacky `time.monotonic()` implementation and no longer provides value.
```
cloud-init mode 'init' took 0.061 seconds (0.06)
```

This is what caused the `(0.06)` part of the above message.

File: cloudinit/cmd/main.py [hunk 20]
```diff
@@ -1273,15 +1277,15 @@ def sub_main(args):
     report_on = True
     if name == "init":
         if args.local:
-            rname, rdesc = ("init-local", "searching for local datasources")
+            rname, rdesc = ("local", "searching for local datasources")
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/cmd/main.py:1280]: I'm fine with keeping the originals for now. The internal references I don't care about as much, but it would benefit users if long-term we got to the point where user-facing references to stages used the same vocabulary as our docs.

I'm sure users find this sequence of logs a bit confusing:
```
2024-05-16 16:59:04,331 - util.py[DEBUG]: cloud-init mode 'init' took 0.362 seconds (0.36)
2024-05-16 16:59:04,331 - handlers.py[DEBUG]: finish: init-local: SUCCESS: searching for local datasources
2024-05-16 16:59:04,778 - util.py[DEBUG]: Cloud-init v. 24.1.3-0ubuntu3 running 'init' at Thu, 16 May 2024 16:59:04 +0000. Up 13.32 seconds.
```

This is confusing:

1. Why did 'init' mode end only to start again?
2. Is init-local different than 'init'?

[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/cmd/main.py]: @TheRealFalcon I dropped this commit for now (out of scope of this PR), but do note that `name` was replaced by `rname` a few lines lower in this file.

**Comment by holmanb** [comment]
@TheRealFalcon I think that I've addressed all of your comments. Ready for re-review (assuming tests pass).

File: cloudinit/performance.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cloudinit/performance.py:40]: I would prefer to allow kwargs here. I think the main case for positional only arguments (other than c api compatibility) is when the arguments are obvious and names really convey no extra meaning. E.g., `min(5, 4)` wouldn't benefit by having kwargs.

However, if we make a call like `Timed(x, threshold=y, log_mode=z)`, there's really no context to tell us what we're actually passing as the first argument. Obviously we should have a better variable name than `x`, but it seems needlessly restricting to disallow the caller to add the context if it'd make the call clearer.

File: cloudinit/performance.py (hunk 15 above)
Comment (TheRealFalcon) [reply cloudinit/performance.py:43]: Ah, thanks. I didn't realize the "skip" option. I go back and forth about enums, but the string option here makes the most sense.

File: cloudinit/config/modules.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cloudinit/config/modules.py]: I'm not seeing the `f"{run_name} ran successfully and took {timer.delta:.3f} seconds"` part.

**Comment by TheRealFalcon** [comment]
@holmanb , See my latest comments on the last two unresolved conversations. +1 to everything else.

File: cloudinit/config/modules.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/config/modules.py]: @TheRealFalcon thanks, just pushed that part

File: cloudinit/performance.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/performance.py]: @TheRealFalcon that seems reasonable, I just removed that restriction

**Comment by holmanb** [comment]
@TheRealFalcon I just pushed fixes for your last two comments

**Review by TheRealFalcon — APPROVED** [review]
LGTM!
